crates/sui-storage/src/event_store/sql.rs

```diff
@@ -0,0 +1,513 @@
+    let tx_digest = digest_raw.map(|bytes| {
+        TransactionDigest::new(
+            bytes
+                .try_into()
+                .expect("Cannot convert digest bytes to TxDigest"),
+        )
+    });
+    let event_type: u16 = row.get(3);
+    let package_id = try_extract_object_id(&row, 4).expect("Error converting package ID bytes");
+    let object_id = try_extract_object_id(&row, 6).expect("Error converting object ID bytes");
+    let module_name: Option<String> = row.get(5);
+    let fields_text: &str = row.get(7);
+    let fields: Vec<_> = if fields_text.is_empty() {
+        Vec::new()
+    } else {
+        let fields_json = serde_json::from_str(fields_text)
+            .unwrap_or_else(|e| panic!("Could not parse [{}] as JSON: {}", fields_text, e));
+        if let Value::Object(map) = fields_json {
+            map.into_iter()
+                .map(|(k, v)| (flexstr::SharedStr::from(k), EventValue::Json(v)))
+                .collect()
+        } else {
+            debug!(?fields_json, "Could not parse JSON as object");
+            Vec::new()
+        }
+    };
+
+    StoredEvent {
+        timestamp: timestamp as u64,
+        checkpoint_num: checkpoint as u64,
+        tx_digest,
+        event_type: SharedStr::from(Event::name_from_ordinal(event_type as usize)),
+        module_name: module_name.map(|s| s.into()),
+        object_id: object_id.or(package_id),
+        fields,
+    }
+}
+
+// Adds JSON fields for items not in any of the standard columns in table definition, eg for MOVE events.
+fn event_to_json(event: &EventEnvelope) -> String {
+    if let Some(json_value) = &event.move_struct_json_value {
+        json_value.to_string()
+    } else {
+        let maybe_json = match &event.event {
+            Event::TransferObject {
+                version,
+                destination_addr,
+                type_,
+                ..
+            } => Some(json!({"destination": destination_addr.to_string(),
+                       "version": version.value(),
+                       "type": type_.to_string() })),
+            _ => None,
+        };
+        maybe_json.map(|j| j.to_string()).unwrap_or_default()
+    }
+}
+
+const SQL_INSERT_TX: &str = "INSERT INTO events (timestamp, checkpoint, tx_digest, event_type, \
+    package_id, module_name, object_id, fields) VALUES (?, ?, ?, ?, ?, ?, ?, ?)";
+
+const TS_QUERY: &str = "SELECT * FROM events WHERE timestamp >= ? AND timestamp < ? LIMIT ?";
+
+const TX_QUERY: &str = "SELECT * FROM events WHERE tx_digest = ?";
+
+const QUERY_BY_TYPE: &str = "SELECT * FROM events WHERE timestamp >= ? AND \
+    timestamp < ? AND event_type = ? ORDER BY timestamp DESC LIMIT ?";
+
+const QUERY_BY_MODULE: &str = "SELECT * FROM events WHERE timestamp >= ? AND \
+    timestamp < ? AND package_id = ? AND module_name = ? ORDER BY timestamp DESC LIMIT ?";
+
+const QUERY_BY_CHECKPOINT: &str =
+    "SELECT * FROM events WHERE checkpoint >= ? AND checkpoint <= ? LIMIT ?";
```

Comment:
does the crate provide parameterized queries or any sanity checks? Although this is not user input, a malicious validator/fullnode can send invalid messages to corrupt a fullnode's database **if there's a vulnerability**. We should be cautious on this front

Reply:
Yes, we are using prepared statements to prevent injection attacks, fairly standard usage here.